docs: add design information around neutron usage (#904)
rackerlabs/understack: docs/design-guide/neutron-networking.md

so to best support this we use the VXLAN type driver with VLANs on all the
leaves bound to the VXLAN VNI. In this configuration there is only one subnet
across the network, in Neutron this is called [L2 Adjacency][neutron-l2-adjacency].
This model has been used by networking vendors such as Arista, Juniper, and Cisco
in their own ML2 mechanism driver. However full support for this has been lacking
upstream so we have developed our own mechanism driver as we explore the best
approach. Traditionally Ironic (baremetal) based Neutron deployment have utilized
the [networking-generic-switch][networking-generic-switch] mechanism however as it
stands today it only supports VLAN and it's templating capabilities are not sufficient
for our needs. We hope to eventually create a generic mechanism which can be contributed
back. Another limitation that we are aware of is in the VXLAN type driver itself,
the Neutron team designed this type as an overlay style VXLAN system while we are
focused on an underlay or EVPN-VXLAN style system. We are actively working with
upstream on how to best include this use case.

## Mapping Leaf/Spine to Neutron Networks

> The Leaf/Spine network is implemented as a VNI that exists on the fabric with
> endpoints on each of the leaves that need connectivity being mapped to a VLAN
> to provide connectivity to physical assets connected to that leaf.

Reviewer: > The Leaf/Spine network is implemented as a VNI

The Leaf/Spine is a physical or logical topology (the underlying network structure for a fabric).
A VNI (VXLAN Network Identifier) is a logical identifier for a specific virtual network segment that runs over the Leaf/Spine fabric (the underlay). You don't implement the topology as a VNI. You implement virtual networks (like VXLAN segments using VNIs) on top of or over the Leaf/Spine topology.

> ...with endpoints on each of the leaves... being mapped to a VLAN

While endpoints (like servers) are connected to ports that are part of a VLAN on the leaf switch, saying the "endpoints... being mapped to a VLAN" is awkward phrasing. More accurately, endpoints reside in or are connected to a VLAN on the access port of the leaf. The VLAN itself is then typically mapped to a VNI at the VTEP (our leaf switch) to transport the traffic across the VXLAN fabric.

should we go with something like:

```suggestion
In a Leaf/Spine fabric, VXLAN VNIs are used to create virtual network segments that run over the IP underlay. On the leaf switches, traditional VLANs connected to physical assets are typically mapped to specific VNIs to provide connectivity across the fabric.
```

### VNIs

To provide available VNIs a [network segment range][neutron-network-segment-range] is

> created of type VXLAN with the name of the fabric.

Author: How about...

An available pool of VNIs is defined by creating a VXLAN network segment range
with the same name as the fabric on which the VNIs will reside.

<!-- markdownlint-capture -->
<!-- markdownlint-disable MD046 -->
!!! Note

    This is one of the places that the underlay vs overlay VXLAN disconnect
    rears its head. Neutron assumes that all VXLAN participants will be able
    to reach each other via the tunnel configuration. However it is possible
    to have multiple physically separate fabrics but Neutron does not allow
    the VXLAN type to have a `physical_network` value like VLAN networks.
<!-- markdownlint-restore -->

Provider networks and self-service networks are allocated VNIs from this range.

### VLANs

For each leaf pair on the fabric a VLAN [network segment range][neutron-network-segment-range]
is created with the `physical_network` value set to their pair name. For example,
the name of the rack they serve could be used.

### Connecting a Server to a Network

When a server needs to have a connection to a network, it's `local_link_connection`
information is looked up. The `physical_network` is then used to attempt to map
the port to how it connects on the network. This is documented in Ironic's

> Networking Guide as [VIF Attachment][ironic-vif-attachment].

Reviewer: When a server needs to establish a connection to a network, Ironic checks the baremetal port's `local_link_connection` and `physical_network` attributes to determine how it should be mapped to the desired network. This information is documented in Ironic's Networking Guide as [VIF Attachment][ironic-vif-attachment].

> If that VNI is not already mapped to a VLAN on the leaf pair where the server

Reviewer: It's unclear where the VNI information originates. We should explain what inputs are used, such as `physical_network` and `local_link_connection`. Additionally, it would be helpful to clarify which component or system is responsible for mapping these attributes to a VNI—whether it's Nautobot, Ironic, or the Neutron driver.

| 78be9792-cf21-4c5e-8432-bd83f0830763 | None             | 783b4239-7220-4a74-8253-415539469860 | vxlan        |  200004 |
+--------------------------------------+------------------+--------------------------------------+--------------+---------+

openstack network segment show 5ab3339d-ae44-4f45-9293-7b41a83bf473
+------------------+--------------------------------------+
| Field            | Value                                |
+------------------+--------------------------------------+
| created_at       | 2025-04-29T13:21:31Z                 |
| description      | None                                 |
| id               | 5ab3339d-ae44-4f45-9293-7b41a83bf473 |
| name             | None                                 |
| network_id       | 783b4239-7220-4a74-8253-415539469860 |
| network_type     | vlan                                 |
| physical_network | f20-2-network                        |
| revision_number  | 0                                    |
| segmentation_id  | 1800                                 |
| updated_at       | 2025-04-29T13:21:31Z                 |
+------------------+--------------------------------------+
```

> Now we can look at the ports and confirm that indeed this segment exists to provide
> connectivity to this server.

Reviewer: ```suggestion
Now we can check the ports to confirm that this segment indeed exists to provide connectivity to the server.
```

| 10099d3c-0ade-41b9-8a1c-1d50ace4bf22 |                 | fa:16:3e:10:8f:f1 | ip_address='192.168.100.1', subnet_id='6f8e3a32-c7a7-4354-808f-75800b21efcf'   | ACTIVE |
| 47bb4c37-f60d-474f-8ce5-c7c1d9982585 | trunk_parent11  | 14:23:f3:f5:22:b0 | ip_address='192.168.100.170', subnet_id='6f8e3a32-c7a7-4354-808f-75800b21efcf' | ACTIVE |
+--------------------------------------+-----------------+-------------------+--------------------------------------------------------------------------------+--------+
```

## ML2 Mechanism Operations

Our ML2 mechanism is responsible for the following:

- creating dynamic VLAN segments on VXLAN networks via port binding operations
- deleting dynamic VLAN segments on VXLAN networks when ports are removed
- triggering the actual operation to update the network devices

### Bind Port

While `bind_port()` is a distinct method inside of an ML2 mechanism, there is
no direct call for this via the Neutron API. This method is triggered by
Neutron based on certain data provided to port creation and update API calls.

> `bind_port()` will be triggered in the following situation:

Reviewer: ```suggestion
`bind_port()` will be triggered in the following situations:
```
